adam/learner/contrastive_learner.py

```diff
@@ -146,7 +146,9 @@ class TeachingContrastiveObjectLearner(ContrastiveLearner):
         validator=instance_of(Counter), factory=Counter, init=False
     )
 
-    def learn_from(self, matching: LanguagePerceptionSemanticContrast) -> None:
+    def learn_from(
+        self, matching: LanguagePerceptionSemanticContrast, top_n: Optional[int] = None
```

Comment:
I prefer to make `top_n` an attribute/property here. I think `top_n` may not make sense for all contrastive learners. It's not required, so this doesn't change the interface, but the script has to pass it in at the point we call learn_from. That means that we have to special case any contrastive learner that doesn't accept this parameter. So it's sort of de facto adding to the contrastive learning interface.

If we use a property it is slightly cleaner to slot a different contrastive learner into the contrastive learning script. We still have to deal with setting `top_n`, but we can do it in a factory function for the contrastive learner where we have to do special casing anyway. See `log_experiment.py`/`build_object_learner_factory()` for example. That contains the specialcasing to our setup so we don't have to worry about those details for the rest of the experiment script.

Reply:
Ok, I've shifted top_n to be an attribute instead of a function parameter.